gb-roms/src/controllers/mbc3.rs

```
@@ -0,0 +1,243 @@
+            0x0000..=0x1FFF => self.regs.ram_enabled = (v & 0xF) == 0xA,
+            0x2000..=0x3FFF => self.regs.rom_bank = if v == 0 { 1 } else { v & 0x7F },
+            0x4000..=0x5FFF => self.regs.ram_bank = v & 0xC,
+            0x6000..=0x7FFF => {
+                if self.regs.last_writed_byte == Some(0_u8) && v == 1 {
+                    self.latch_clock_data(addr)?;
+                } else {
+                    self.regs.last_writed_byte = Some(v);
+                }
+            }
+            _ => return Err(Error::new_segfault(addr)),
+        }
+        Ok(())
+    }
+
+    fn latch_clock_data(&mut self, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        if let Some(clock) = self.clock.as_ref() {
+            self.regs.rtc = clock.into();
+        } else {
+            return Err(Error::new_segfault(addr));
+        }
+        Ok(())
+    }
+
+    fn read_ram(&self, addr: Box<dyn Address<Area>>) -> Result<u8, Error> {
+        if !self.regs.ram_enabled {
+            return Err(Error::new_segfault(addr));
+        }
+        let address = addr.get_address();
+        let ram_bank = self.regs.ram_bank;
+        match ram_bank {
+            0x0..=0x3 => Ok(self.ram_banks[ram_bank as usize][(address as usize) & 0x1FFF]),
+            0x8 => Ok(self.regs.rtc.seconds),
+            0x9 => Ok(self.regs.rtc.minutes),
+            0xA => Ok(self.regs.rtc.hours),
+            0xB => Ok(self.regs.rtc.lower_day_counter),
+            0xC => Ok(self.regs.rtc.upper_day_counter),
+            _ => Err(Error::new_segfault(addr)),
+        }
+    }
+
+    fn write_ram(&mut self, v: u8, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        if !self.regs.ram_enabled {
+            return Err(Error::new_segfault(addr));
+        }
+        let address = addr.get_address();
+        let ram_bank = self.regs.ram_bank;
+        match ram_bank {
+            0x0..=0x3 => self.ram_banks[ram_bank as usize][(address as usize) & 0x1FFF] = v,
```

remove the bitwise on the `address` it hide potential error

---

Even if it could provoke  a crash ?

---

and it the role the `gb-bus` to check the mapped area